File: grimorio de estudos/Cs50/Problem Sets/sentimental-credit/credit.py
from cs50 import get_int

def main():

        cardNumber = get_int("Please input your credit card number: ")

        if cardValidation(cardNumber):
            cardCompany(cardNumber)
        else:
            print("INVALID")


def cardValidation(input):
    # Calculates the sum using Luhn's Algorithm
    totalSum = 0
    for x, y in enumerate(reversed(str(input))):
        if x % 2 == 0:
            totalSum += int(y)
        else:
            for z in str(int(y) * 2):
                totalSum += int(z)

    # Checks if sum is % by 10
    if totalSum % 10 == 0:
        return True
    else:
        #print("INVALID")
        return False

def cardCompany(input):
    # Gets first two numbers of input
    number = int(str(input)[0:2])

    # Checks for known company
    if number == 34 or number == 37:
        print("AMEX")
    if number >= 40 and number < 50:
        print("VISA")
    if number > 50 and number < 56:
        print("MASTERCARD")
    # Else there is no company, INVALID



if __name__ == "__main__":
    main()

File: harvardCs50/2022/Problem Sets/credit/credit.c
#include <cs50.h>
#include <stdio.h>

//it was really hard to wrap my head around this, is this really the first problem set?
int main(void){

    //number input
    long creditNumber = get_long("Please input your credit card number: ");
    printf("Credit Number: %.0li\n", creditNumber);

    //check number length
    int length = 0;
    long n = creditNumber;

    while (n > 0)
    {
        n = n / 10;
        length++;
    }
    if (length != 13 && length != 15 && length != 16)
    {
        printf("INVALID\n");
        return 0;
    }clear


    //calculate the sum
    int firstSum = 0;
    int secondSum = 0;
    long x = creditNumber;
    int total = 0;
    int mod1;
    int mod2;
    int d1;
    int d2;
    do
    {
        // Remove last digit and add to firstSum
        mod1 = x % 10;
        x = x / 10;
        firstSum = firstSum + mod1;

        // Remove second last digit
        mod2 = x % 10;
        x = x / 10;

        // Double second last digit and add digits to secondSum
        mod2 = mod2 * 2;
        d1 = mod2 % 10;
        d2 = mod2 / 10;
        secondSum = secondSum + d1 + d2;

    }
    while (x > 0);
    total = firstSum + secondSum;
    //next check Luhn Algorithm
    if (total % 10 != 0)
    {
        printf("INVALID\n");
        return 0;
    }
    //get starting digits
    long start = creditNumber;
    do
    {
        start = start / 10;
    }
    while (start > 100);
    //check starting digits for card type
    //mastercard
    if ((start / 10 == 5) && (0 < start % 10 && start % 10 < 6))
    {
        printf("MASTERCARD\n");
    }
    else if ((start / 10 == 3) && (start % 10 == 4 || start % 10 == 7))
    {
        printf("AMEX\n");
    }
    else if (start / 10 == 4)
    {
        printf("VISA\n");
    }
    else
    {
        printf("INVALID\n");
    }
}
